pkg/gce-pd-csi-driver/controller_test.go

```diff
@@ -1413,3 +1413,76 @@ func TestPickRandAndConsecutive(t *testing.T) {
 
 	}
 }
+
+func TestVolumeOperationConcurrency(t *testing.T) {
+	snapshotToExecute := make(chan gce.SnapshotSourceAndTarget, 3)
+	readyToExecute := make(chan struct{}, 2)
+	reqFinished := make(chan error, 2)
+
+	gceDriver := initBlockingGCEDriver(t, nil, snapshotToExecute, readyToExecute)
+	cs := gceDriver.cs
+
+	vol1CreateSnapshotReqA := &csi.CreateSnapshotRequest{
+		Name:           name + "1A",
+		SourceVolumeId: testVolumeId + "1",
+	}
+	vol1CreateSnapshotReqB := &csi.CreateSnapshotRequest{
+		Name:           name + "1B",
+		SourceVolumeId: testVolumeId + "1",
+	}
+	vol2CreateSnapshotReq := &csi.CreateSnapshotRequest{
+		Name:           name + "2",
+		SourceVolumeId: testVolumeId + "2",
+	}
+
+	runRequestInBackground := func(req *csi.CreateSnapshotRequest) {
+		_, err := cs.CreateSnapshot(context.Background(), req)
+		reqFinished <- err
+	}
+
+	sourceAndTargetForRequest := func(req *csi.CreateSnapshotRequest) gce.SnapshotSourceAndTarget {
+		volKey, err := common.VolumeIDToKey(req.SourceVolumeId)
+		if err != nil {
+			t.Fatalf("Could not find volume with ID %v: %v", req.SourceVolumeId, err)
+		}
+		return gce.SnapshotSourceAndTarget{
+			Source: volKey.String(),
+			Target: req.Name,
+		}
+	}
+
+	// Start first valid request vol1CreateSnapshotReqA and block until it reaches the CreateSnapshot
+	go runRequestInBackground(vol1CreateSnapshotReqA)
+	<-readyToExecute
+
+	// Check that vol1CreateSnapshotReqB is rejected, due to same volume ID
+	// Also allow vol1CreateSnapshotReqB to complete, in case it is allowed to Mount
+	snapshotToExecute <- sourceAndTargetForRequest(vol1CreateSnapshotReqB)
```

Comment:
The pattern you are using here can be described as, "Signal on a channel once the request has begun, and wait for a return signal before allowing the request to complete."

I like this pattern.

A lot of the complexity in this test comes from the fact that you are using two independent channels to do this. An alternative is to use a channel of channels. That is, each request starts by creating a reply channel and sending it to the unit test through the "main" channel. Then each request waits for a reply on its own channel, and the test has clearer control over which request is allowed to proceed.

Reply:
I just tried implementing this, but it actually don't because of how the call is set up. The test interface blocks in the cloudprovider function call, while logic that aborts the operation occurs in the layer above that. So if the operation is aborted, it will never make it to the cloud provider call. This means that the channel of channels won't get a chance to create and send the reply channel.

Reply:
The problem arises from the fact that merging the two independent channels means that the caller will need to block on the "main" channel in order to confirm that the operation has started. In the current setup, the operation that should be aborted has the "go ahead" SnapshotSourceAndTarget prequeued so that we can directly just go and check the error channel. In this set up, in order to give the go ahead we need the reply channel that is generated by the function call, but because that call is aborted, the reply channel isn't created.

Reply:
I think you can still cause the first call to block while the second call gets an abort error. You don't need to block or unblock any threads for that. I'm a little sleep deprived, so maybe not understanding the concurrency correctly.